File: archive/cdktf-ts/Pr665/lib/constructs/security.ts
import { Construct } from 'constructs';
import {
  securityGroup,
  securityGroupRule,
  iamRole,
  iamRolePolicy,
  iamInstanceProfile,
  dataAwsIamPolicyDocument,
  ssmParameter,
  wafv2WebAcl,
  wafv2IpSet,
} from '@cdktf/provider-aws/lib';
import { AppConfig } from '../config/variables';

export interface SecurityProps {
  config: AppConfig;
  vpcId: string;
}

export class SecurityConstruct extends Construct {
  public readonly albSecurityGroup: securityGroup.SecurityGroup;
  public readonly ec2SecurityGroup: securityGroup.SecurityGroup;
  public readonly rdsSecurityGroup: securityGroup.SecurityGroup;
  public readonly ec2Role: iamRole.IamRole;
  public readonly ec2InstanceProfile: iamInstanceProfile.IamInstanceProfile;
  public readonly webAcl: wafv2WebAcl.Wafv2WebAcl;

  constructor(scope: Construct, id: string, props: SecurityProps) {
    super(scope, id);

    const { config, vpcId } = props;

    this.albSecurityGroup = new securityGroup.SecurityGroup(this, 'alb-sg', {
      name: `${config.projectName}-${config.environment}-alb-sg`,
      description: 'Security group for Application Load Balancer',
      vpcId: vpcId,
      tags: {
        ...config.tags,
        Name: `${config.projectName}-alb-sg`,
      },
    });

    new securityGroupRule.SecurityGroupRule(this, 'alb-http-ingress', {
      type: 'ingress',
      fromPort: 80,
      toPort: 80,
      protocol: 'tcp',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: this.albSecurityGroup.id,
      description: 'Allow HTTP traffic from internet',
    });

    new securityGroupRule.SecurityGroupRule(this, 'alb-https-ingress', {
      type: 'ingress',
      fromPort: 443,
      toPort: 443,
      protocol: 'tcp',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: this.albSecurityGroup.id,
      description: 'Allow HTTPS traffic from internet',
    });

    new securityGroupRule.SecurityGroupRule(this, 'alb-egress', {
      type: 'egress',
      fromPort: 0,
      toPort: 65535,
      protocol: 'tcp',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: this.albSecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    this.ec2SecurityGroup = new securityGroup.SecurityGroup(this, 'ec2-sg', {
      name: `${config.projectName}-${config.environment}-ec2-sg`,
      description: 'Security group for EC2 instances',
      vpcId: vpcId,
      tags: {
        ...config.tags,
        Name: `${config.projectName}-ec2-sg`,
      },
    });

    new securityGroupRule.SecurityGroupRule(this, 'ec2-alb-ingress', {
      type: 'ingress',
      fromPort: 80,
      toPort: 80,
      protocol: 'tcp',
      sourceSecurityGroupId: this.albSecurityGroup.id,
      securityGroupId: this.ec2SecurityGroup.id,
      description: 'Allow HTTP traffic from ALB',
    });

    new securityGroupRule.SecurityGroupRule(this, 'ec2-ssh-ingress', {
      type: 'ingress',
      fromPort: 22,
      toPort: 22,
      protocol: 'tcp',
      cidrBlocks: ['10.0.0.0/16'],
      securityGroupId: this.ec2SecurityGroup.id,
      description: 'Allow SSH from VPC',
    });

    new securityGroupRule.SecurityGroupRule(this, 'ec2-egress', {
      type: 'egress',
      fromPort: 0,
      toPort: 65535,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: this.ec2SecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    this.rdsSecurityGroup = new securityGroup.SecurityGroup(this, 'rds-sg', {
      name: `${config.projectName}-${config.environment}-rds-sg`,
      description: 'Security group for RDS database',
      vpcId: vpcId,
      tags: {
        ...config.tags,
        Name: `${config.projectName}-rds-sg`,
      },
    });

    new securityGroupRule.SecurityGroupRule(this, 'rds-ec2-ingress', {
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      sourceSecurityGroupId: this.ec2SecurityGroup.id,
      securityGroupId: this.rdsSecurityGroup.id,
      description: 'Allow PostgreSQL access from EC2 instances',
    });

    this.ec2Role = new iamRole.IamRole(this, 'ec2-role', {
      name: `${config.projectName}-${config.environment}-ec2-role`,
      assumeRolePolicy: new dataAwsIamPolicyDocument.DataAwsIamPolicyDocument(
        this,
        'ec2-assume-role-policy',
        {
          statement: [
            {
              actions: ['sts:AssumeRole'],
              principals: [
                {
                  type: 'Service',
                  identifiers: ['ec2.amazonaws.com'],
                },
              ],
            },
          ],
        }
      ).json,
      tags: config.tags,
    });

    new iamRolePolicy.IamRolePolicy(this, 'ec2-policy', {
      name: `${config.projectName}-${config.environment}-ec2-policy`,
      role: this.ec2Role.id,
      policy: new dataAwsIamPolicyDocument.DataAwsIamPolicyDocument(
        this,
        'ec2-policy-document',
        {
          statement: [
            {
              actions: ['s3:GetObject', 's3:PutObject', 's3:DeleteObject'],
              resources: [
                `arn:aws:s3:::${config.projectName}-${config.environment}-logs-*/*`,
              ],
            },
            {
              actions: ['s3:ListBucket'],
              resources: [
                `arn:aws:s3:::${config.projectName}-${config.environment}-logs-*`,
              ],
            },
            {
              actions: [
                'ssm:GetParameter',
                'ssm:GetParameters',
                'ssm:GetParametersByPath',
              ],
              resources: [
                `arn:aws:ssm:${config.region}:*:parameter/${config.projectName}/*`,
              ],
            },
            {
              actions: [
                'cloudwatch:PutMetricData',
                'logs:PutLogEvents',
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
              ],
              resources: ['*'],
            },
          ],
        }
      ).json,
    });

    this.ec2InstanceProfile = new iamInstanceProfile.IamInstanceProfile(
      this,
      'ec2-instance-profile',
      {
        name: `${config.projectName}-${config.environment}-ec2-instance-profile`,
        role: this.ec2Role.name,
        tags: config.tags,
      }
    );

    new ssmParameter.SsmParameter(this, 'db-host', {
      name: `/${config.projectName}/database/host`,
      type: 'String',
      value: 'placeholder',
      description: 'Database host endpoint',
      tags: config.tags,
    });

    new ssmParameter.SsmParameter(this, 'db-name', {
      name: `/${config.projectName}/database/name`,
      type: 'String',
      value: 'webapp',
      description: 'Database name',
      tags: config.tags,
    });

    new ssmParameter.SsmParameter(this, 'db-username', {
      name: `/${config.projectName}/database/username`,
      type: 'SecureString',
      value: 'webapp_user',
      description: 'Database username',
      tags: config.tags,
    });

    new ssmParameter.SsmParameter(this, 'db-password', {
      name: `/${config.projectName}/database/password`,
      type: 'SecureString',
      value: 'ChangeMe123!',
      description: 'Database password',
      tags: config.tags,
    });

    const ipSet = new wafv2IpSet.Wafv2IpSet(this, 'blocked-ips', {
      name: `${config.projectName}-${config.environment}-blocked-ips`,
      scope: 'REGIONAL',
      ipAddressVersion: 'IPV4',
      addresses: ['192.0.2.44/32'],
      description: 'Blocked IP addresses',
      tags: config.tags,
    });

    this.webAcl = new wafv2WebAcl.Wafv2WebAcl(this, 'web-acl', {
      name: `${config.projectName}-${config.environment}-web-acl`,
      scope: 'REGIONAL',
      description: 'WAF Web ACL for application protection',
      defaultAction: {
        allow: {},
      },
      rule: [
        {
          name: 'BlockedIPs',
          priority: 1,
          action: {
            block: {},
          },
          statement: {
            ipSetReferenceStatement: {
              arn: ipSet.arn,
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: 'BlockedIPsRule',
            sampledRequestsEnabled: true,
          },
        },
        {
          name: 'AWSManagedRulesCommonRuleSet',
          priority: 2,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: 'AWSManagedRulesCommonRuleSet',
              vendorName: 'AWS',
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: 'CommonRuleSetMetric',
            sampledRequestsEnabled: true,
          },
        },
        {
          name: 'AWSManagedRulesKnownBadInputsRuleSet',
          priority: 3,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: 'AWSManagedRulesKnownBadInputsRuleSet',
              vendorName: 'AWS',
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: 'KnownBadInputsRuleSetMetric',
            sampledRequestsEnabled: true,
          },
        },
        {
          name: 'AWSManagedRulesSQLiRuleSet',
          priority: 4,
          overrideAction: {
            none: {},
          },
          statement: {
            managedRuleGroupStatement: {
              name: 'AWSManagedRulesSQLiRuleSet',
              vendorName: 'AWS',
            },
          },
          visibilityConfig: {
            cloudwatchMetricsEnabled: true,
            metricName: 'SQLiRuleSetMetric',
            sampledRequestsEnabled: true,
          },
        },
      ],
      tags: config.tags,
      visibilityConfig: {
        cloudwatchMetricsEnabled: true,
        metricName: `${config.projectName}WebAcl`,
        sampledRequestsEnabled: true,
      },
    });

    this.webAcl.addOverride('rule.0.statement.ip_set_reference_statement', {
      arn: ipSet.arn,
    });
    this.webAcl.addOverride(
      'rule.0.statement.ipSetReferenceStatement',
      undefined
    );

    this.webAcl.addOverride('rule.1.statement.managed_rule_group_statement', {
      name: 'AWSManagedRulesCommonRuleSet',
      vendor_name: 'AWS',
    });
    this.webAcl.addOverride(
      'rule.1.statement.managedRuleGroupStatement',
      undefined
    );

    this.webAcl.addOverride('rule.2.statement.managed_rule_group_statement', {
      name: 'AWSManagedRulesKnownBadInputsRuleSet',
      vendor_name: 'AWS',
    });
    this.webAcl.addOverride(
      'rule.2.statement.managedRuleGroupStatement',
      undefined
    );

    this.webAcl.addOverride('rule.3.statement.managed_rule_group_statement', {
      name: 'AWSManagedRulesSQLiRuleSet',
      vendor_name: 'AWS',
    });
    this.webAcl.addOverride(
      'rule.3.statement.managedRuleGroupStatement',
      undefined
    );
  }
}

File: archive/cdktf-py/Pr6460/lib/security.py
"""Security infrastructure module with IAM, security groups, and WAF."""

from constructs import Construct
from cdktf_cdktf_provider_aws.security_group import (
    SecurityGroup,
    SecurityGroupIngress,
    SecurityGroupEgress,
)
from cdktf_cdktf_provider_aws.iam_role import IamRole
from cdktf_cdktf_provider_aws.iam_role_policy import IamRolePolicy
from cdktf_cdktf_provider_aws.iam_instance_profile import IamInstanceProfile
from cdktf_cdktf_provider_aws.wafv2_web_acl import Wafv2WebAcl
from cdktf import TerraformResource
import json


class SecurityInfrastructure(Construct):
    """Security infrastructure with IAM roles, security groups, and WAF."""

    def __init__(self, scope: Construct, construct_id: str, environment_suffix: str, vpc_id: str):
        """
        Initialize security infrastructure.

        Args:
            scope: The scope in which to define this construct
            construct_id: The scoped construct ID
            environment_suffix: Unique suffix for resource naming
            vpc_id: VPC ID for security groups
        """
        super().__init__(scope, construct_id)

        # ALB Security Group
        self.alb_sg = SecurityGroup(
            self,
            "alb_sg",
            name=f"payment-alb-sg-{environment_suffix}",
            description="Security group for Application Load Balancer",
            vpc_id=vpc_id,
            ingress=[
                SecurityGroupIngress(
                    from_port=443,
                    to_port=443,
                    protocol="tcp",
                    cidr_blocks=["0.0.0.0/0"],
                    description="HTTPS from internet",
                ),
                SecurityGroupIngress(
                    from_port=80,
                    to_port=80,
                    protocol="tcp",
                    cidr_blocks=["0.0.0.0/0"],
                    description="HTTP from internet (redirect to HTTPS)",
                ),
            ],
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"],
                    description="All outbound traffic",
                )
            ],
            tags={
                "Name": f"payment-alb-sg-{environment_suffix}",
            },
        )

        # Application Security Group
        self.app_sg = SecurityGroup(
            self,
            "app_sg",
            name=f"payment-app-sg-{environment_suffix}",
            description="Security group for application instances",
            vpc_id=vpc_id,
            ingress=[
                SecurityGroupIngress(
                    from_port=8080,
                    to_port=8080,
                    protocol="tcp",
                    security_groups=[self.alb_sg.id],
                    description="Application port from ALB",
                )
            ],
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"],
                    description="All outbound traffic",
                )
            ],
            tags={
                "Name": f"payment-app-sg-{environment_suffix}",
            },
        )

        # Database Security Group
        self.db_sg = SecurityGroup(
            self,
            "db_sg",
            name=f"payment-db-sg-{environment_suffix}",
            description="Security group for RDS database",
            vpc_id=vpc_id,
            ingress=[
                SecurityGroupIngress(
                    from_port=5432,
                    to_port=5432,
                    protocol="tcp",
                    security_groups=[self.app_sg.id],
                    description="PostgreSQL from application",
                )
            ],
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"],
                    description="All outbound traffic",
                )
            ],
            tags={
                "Name": f"payment-db-sg-{environment_suffix}",
            },
        )

        # IAM Role for EC2 instances
        assume_role_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {"Service": "ec2.amazonaws.com"},
                    "Action": "sts:AssumeRole",
                }
            ],
        }

        self.instance_role = IamRole(
            self,
            "instance_role",
            name=f"payment-app-role-{environment_suffix}",
            assume_role_policy=json.dumps(assume_role_policy),
            tags={
                "Name": f"payment-app-role-{environment_suffix}",
            },
        )

        # IAM Policy for EC2 instances (least privilege)
        # S3 bucket name pattern: payment-static-{environment_suffix}
        s3_bucket_pattern = f"payment-static-{environment_suffix}"
        instance_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": ["s3:ListBucket"],
                    "Resource": [f"arn:aws:s3:::{s3_bucket_pattern}"],
                },
                {
                    "Effect": "Allow",
                    "Action": ["s3:GetObject"],
                    "Resource": [f"arn:aws:s3:::{s3_bucket_pattern}/*"],
                },
                {
                    "Effect": "Allow",
                    "Action": [
                        "cloudwatch:PutMetricData",
                        "logs:CreateLogGroup",
                        "logs:CreateLogStream",
                        "logs:PutLogEvents",
                    ],
                    "Resource": "*",
                },
                {
                    "Effect": "Allow",
                    "Action": [
                        "ssm:GetParameter",
                        "ssm:GetParameters",
                    ],
                    "Resource": f"arn:aws:ssm:*:*:parameter/payment-processing/{environment_suffix}/*",
                },
            ],
        }

        IamRolePolicy(
            self,
            "instance_policy",
            name=f"payment-app-policy-{environment_suffix}",
            role=self.instance_role.id,
            policy=json.dumps(instance_policy),
        )

        # Instance Profile
        self.instance_profile = IamInstanceProfile(
            self,
            "instance_profile",
            name=f"payment-app-profile-{environment_suffix}",
            role=self.instance_role.name,
        )

        # WAF Web ACL with AWS Managed Rule Groups
        # Using raw Terraform resource due to CDKTF Python limitations with nested structures
        self.waf_acl = TerraformResource(
            self,
            "waf_acl",
            terraform_resource_type="aws_wafv2_web_acl"
        )

        # Add WAF configuration using overrides
        self.waf_acl.add_override("name", f"payment-waf-{environment_suffix}")
        self.waf_acl.add_override("scope", "REGIONAL")
        self.waf_acl.add_override("default_action", {"allow": {}})

        # Add managed rule groups
        self.waf_acl.add_override("rule", [
            {
                "name": "AWSManagedRulesCommonRuleSet",
                "priority": 1,
                "statement": {
                    "managed_rule_group_statement": {
                        "vendor_name": "AWS",
                        "name": "AWSManagedRulesCommonRuleSet"
                    }
                },
                "override_action": {
                    "none": {}
                },
                "visibility_config": {
                    "cloudwatch_metrics_enabled": True,
                    "metric_name": "AWSManagedRulesCommonRuleSetMetric",
                    "sampled_requests_enabled": True
                }
            },
            {
                "name": "AWSManagedRulesKnownBadInputsRuleSet",
                "priority": 2,
                "statement": {
                    "managed_rule_group_statement": {
                        "vendor_name": "AWS",
                        "name": "AWSManagedRulesKnownBadInputsRuleSet"
                    }
                },
                "override_action": {
                    "none": {}
                },
                "visibility_config": {
                    "cloudwatch_metrics_enabled": True,
                    "metric_name": "AWSManagedRulesKnownBadInputsRuleSetMetric",
                    "sampled_requests_enabled": True
                }
            },
            {
                "name": "AWSManagedRulesSQLiRuleSet",
                "priority": 3,
                "statement": {
                    "managed_rule_group_statement": {
                        "vendor_name": "AWS",
                        "name": "AWSManagedRulesSQLiRuleSet"
                    }
                },
                "override_action": {
                    "none": {}
                },
                "visibility_config": {
                    "cloudwatch_metrics_enabled": True,
                    "metric_name": "AWSManagedRulesSQLiRuleSetMetric",
                    "sampled_requests_enabled": True
                }
            }
        ])

        # Add visibility config and tags
        self.waf_acl.add_override("visibility_config", {
            "cloudwatch_metrics_enabled": True,
            "metric_name": f"payment-waf-{environment_suffix}",
            "sampled_requests_enabled": True
        })
        self.waf_acl.add_override("tags", {
            "Name": f"payment-waf-{environment_suffix}"
        })

    @property
    def alb_security_group_id(self) -> str:
        """Return ALB security group ID."""
        return self.alb_sg.id

    @property
    def app_security_group_id(self) -> str:
        """Return application security group ID."""
        return self.app_sg.id

    @property
    def db_security_group_id(self) -> str:
        """Return database security group ID."""
        return self.db_sg.id

    @property
    def instance_profile_name(self) -> str:
        """Return instance profile name."""
        return self.instance_profile.name

    @property
    def waf_web_acl_arn(self) -> str:
        """Return WAF Web ACL ARN."""
        return self.waf_acl.get_string_attribute("arn")
